### ISSUE TITLE
TLS impossible state with Pomerium Istio multi-cluster setup

### ISSUE THREAD

**Issue by eddiewang** [pomerium/pomerium#2386]
Currently I have a 2 cluster setup. Both are running Istio mode, using the guide [here](https://www.pomerium.io/guides/istio.html#istio)

The primary cluster holds all 4 services: authz, authn, databroker, and proxy. It's set to insecure mode, with Istio sidecars managing the TLS between services. This works well, and I've been able to expose services in this cluster. TLS is covered and managed by LetsEncrypt, and so we can expose `https://databroker.proxy.company.com` as a secure GRPC service for the secondary cluster to consume.

The secondary cluster only hosts the authz and proxy services. This deployment is also using Istio's service-mesh, and so is set to insecure mode. The authz and proxy service can talk to each other, but cannot talk to the databroker. We can fix this issue by purposely setting GRPC_INSECURE mode to false, but then authz and proxy can no longer talk to themselves.

Problem is, the proxy service needs to connect with both the databroker service, as well as the authorize service. Since we have to set the insecure flag in Istio-mode, we need a way to signal to Pomerium that the databroker service is a secure GRPC service, while the authorize service is meant to be insecure.

Right now it seems like you can only flip one switch, which becomes a global setting for all of the service clients.

**Comment by calebdoxsey** [comment]
Hi Eddie,

Currently we don't support customizing the `GRPC_INSECURE` option by service. There are a few workarounds you could try:

1. If you're using redis for the databroker, you can have multiple databrokers backed by the same redis cluster. So you could have a databroker for each cluster rather than using a single databroker
2. You can turn on TLS for the gRPC connections. I haven't tried this with istio, but there's probably way to do it. (it would mean using some sort of local cert, or just trusting anything)

I do think it would be useful to support per-service customization of the gRPC TLS settings. I worry about breaking existing configuration though. Since the gRPC addresses are specified as URLs it seems like we should be able to leverage the scheme to determine whether or not to use TLS. e.g.:

```
authorize_service_url: http://localhost:5443
databroker_service_url: https://notlocalhost:5443
```

Theoretically we could support a configuration like that to turn TLS on/off via `http`/`https` in the URL.

The concern would be if there are existing URLs out there defined as `https://...` but using `grpc_insecure`, which would then break if started relying on the scheme. But maybe that's unusual and we just bite the bullet? We will need to think about it.

**Comment by travisgroth** [comment]
We've been updating our deployment guide for Istio but I'm not sure there's any concrete solution in the new configuration recommendations.

@eddiewang would it be practical to use istio between the clusters rather than going outside the mesh?

**Comment by eddiewang** [comment]
@travisgroth do you mind explaining what that means?

current topology is as follows:
Main cluster - whole helm stack is installed
Child cluster - Authorize and Proxy is installed

challenge:
- When using istio, we have to use all sorts of tricky SNI stuff to get this all wired up happily
- We are forced to turn on secure mode in Istio because without it on, the child cluster services cannot reach the main cluster.
- To be clear, the main cluster is able to function properly with insecure mode on (as this is the standard install), only child clusters have wonky setup processes.

**Comment by travisgroth** [comment]
> @travisgroth do you mind explaining what that means?

It sounds like you have two clusters with somewhat independent meshes.  My understanding is there are several models for mesh federation in istio, but you should be able to route traffic to the main cluster's databroker using the service mesh for TLS.  Once you have some federation established, I _believe_ this is facilitated with a CRD for the service in the child cluster pointed at an appropriately configured IngressGateway but I haven't looked into the topic in a while.  This should allow you to run all Pomerium components in insecure mode since Istio is handling all TLS.

There's still room for improvement to make Pomerium's TLS configuration more granular, but I wanted to suggest a possible solution in the mean time.

**Comment by eddiewang** [comment]
Ah yes, Istio multicluster does exist and is a possible solution, but in our deployment model it creates more complexity than solutions, so we don't opt-in for that feature. I appreciate the suggestion though!

Fwiw we are able to get this independant mesh model to work with Pomerium as is. It just requires some trickery with SNI routing that make the experience substandard. The idea solution here would be the ability to independently configure secure / insecure mode depending on the service.

**Comment by travisgroth** [comment]
> Fwiw we are able to get this independant mesh model to work with Pomerium as is. It just requires some trickery with SNI routing that make the experience substandard. The idea solution here would be the ability to independently configure secure / insecure mode depending on the service.

👍 yes understood.

**Comment by travisgroth** [comment]
Just to restate @calebdoxsey's proposal - I think using the URI to specify the scheme for clients makes sense.  Though it would be a breaking change, it seems worth making for the next minor release.  I suspect this won't impact many users.

**Comment by travisgroth** [comment]
@eddiewang it looks like this was addressed already as part of #2469 and should be on v0.16+.
